### PR TITLE
rfc: dropout primitive attribute

### PR DESCRIPTION
**Pull request by itaraban**
This is proposal to support dropout operation in oneDNN via primitive attribute
[Link ](https://github.com/itaraban/oneDNN/blob/itaraban/rfcs/dropout/rfcs/20230818-Dropout/README.md)to rendered document
[Link](https://github.com/itaraban/oneDNN/tree/itaraban/dropout-impl) to PoC
Performance data for [DGL](https://github.com/dmlc/dgl) (with PyTorch backend) GNN training benchmarks using Icelake server machine:
| benchmark    | dataset       | default     | with oneDNN dropout | difference |
|--------------|---------------|-------------|---------------------|------------|
| bench_gat    |  cora         | 0.012357805 | 0.009989643         | 124%       |
|              | pubmed        | 0.029577794 | 0.021833464         | 135%       |
| bench_sage   | cora          | 0.007244702 | 0.00432033         | 168%       |
|              | pubmed        | 0.017254775 | 0.009778418         | 176%       |

### CODE REVIEW

File: rfcs/20230818-Dropout/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+```c
+/// Returns probability for drop-out primitive attribute.
+///
+/// @param attr Primitive attributes.
+/// @param p Drop-out probability
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t DNNL_API dnnl_primitive_attr_get_dropout(
+        const_dnnl_primitive_attr_t attr,
+        float *p, const_dnnl_memory_desc_t *drop_desc);
```
[the remark this answers is not in the document]
Comment (itaraban) [reply rfcs/20230818-Dropout/README.md:22]: Yes, mentioned algorithm is what we expect from this primitive attiribute - compute and apply mask.
On backward pass mask from forward will be multiplied by input(see [native_dropout_backward from PyTorch](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/Dropout.cpp#L130))
The main idea was to move mask generation inside of oneDNN. Otherwise user can just use binary_mul post op to apply existing mask.

File: rfcs/20230818-Dropout/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+/// Sets probability for drop-out primitive attribute.
+///
+/// @param attr Primitive attributes.
+/// @param p Drop-out probability
```
[the remark this answers is not in the document]
Comment (itaraban) [reply rfcs/20230818-Dropout/README.md:27]: In our models/benchmarks we have same probability for all layers, but I think, that it is possible to make this runtime parameter.

File: rfcs/20230818-Dropout/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mgouicem) [reply rfcs/20230818-Dropout/README.md:22]: Thanks. That clarifies the scope of the rfc (the new attribute is for fwd only).

[the remark this answers is not in the document]
Comment (mgouicem) [reply rfcs/20230818-Dropout/README.md:22]: BTW, what datatypes is oneDNN expected to support for the mask?

**Comment by itaraban** [comment]
> From the POC, it seems that in ref path the dropout is applied after post-ops and not before. Is that intentional?

No, I'll change it. Thank you for finding it!

File: rfcs/20230818-Dropout/README.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+```c
+/// Returns the parameters of a drop out attribute.
+///
+/// @param attr Primitive attributes.
+/// @param enable_drop Drop-out enable/disable flag
+/// @param mask_desc Output memory descriptor of a drop out mask.
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t DNNL_API dnnl_primitive_attr_get_dropout(
```
Comment (igorsafo) [inline rfcs/20230818-Dropout/README.md:21]: Why is Droppout added as an attribute and not a post operation?
For post operations there is more flexibility in case we need to adjust order in which operations are applied after the main operation.

File: rfcs/20230818-Dropout/README.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+/// Set up drop-out primitive attribute.
+///
+/// @param attr Primitive attributes.
+/// @param enable_drop Drop-out enable/disable flag
+/// @param mask_desc Output memory descriptor of a drop out mask.
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t DNNL_API dnnl_primitive_attr_set_dropout(
+        dnnl_primitive_attr_t attr, uint8_t enable_drop,
```
Comment (igorsafo) [inline rfcs/20230818-Dropout/README.md:33]: I would suggest to remove `enable_drop` from both APIs:
- In getter rely on md returned from the API call. If it is a zero md, then dropout is not set by user.
- In setter -- once user called `dnnl_primitive_attr_set_dropout(attr, mask)` dropout is set.

The current API is a bit confusing when user calls `dnnl_primitive_attr_set_dropout(attr, **false**, mask_desc)`

File: rfcs/20230818-Dropout/README.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+/// @param mask_desc Output memory descriptor of a drop out mask.
+void set_dropout(bool enabled, const memory::desc &mask_desc) {
+    error::wrap_c_api(dnnl_primitive_attr_set_dropout(get(), enabled, mask_desc.get()),
+            "could not set dropout primitive attribute");
+}
+```
+and runtime dropout arguments: output mask, which can be used in backward pass,
+dropout probability and seed.
+
+```c
```
Comment (igorsafo) [inline rfcs/20230818-Dropout/README.md:68]: (random spot): We might need to introduce a standalone Dropout primitive to support frrameworks like ONNX that registers operations supported by the backend. In the current proposal Dropout functionality will be limited to a few patterns, but the rest patterns will not be able to implement Dropout functionality using oneDNN. Please double check with the frameworks if the solution works for them as well.
+@georgen117

File: rfcs/20230818-Dropout/README.md (hunk 3 above)
Comment (mgouicem) [reply rfcs/20230818-Dropout/README.md:21]: The suggestion to use attribute mainly comes from complex primitives, like RNN, where post-op semantics is not well defined, and for which we already had requests in the past for fused dropout.

File: rfcs/20230818-Dropout/README.md (hunk 5 above)
Comment (mgouicem) [reply rfcs/20230818-Dropout/README.md:68]: I guess the question here is "is there a benefit to support dropout primitive in oneDNN"?
For the case of ONNX, isn't it ok if the oneDNN provider provides the dropout operation without using oneDNN, but uses oneDNN when dropout fusion occurs?

File: rfcs/20230818-Dropout/README.md (hunk 3 above)
Comment (igorsafo) [reply rfcs/20230818-Dropout/README.md:21]: > The suggestion to use attribute mainly comes from complex primitives, like RNN, where post-op semantics is not well defined, and for which we already had requests in the past for fused dropout.

O, RNN... Makes sense. @itaraban Could you please add this information somewhere in the RFC?

Comment (itaraban) [reply rfcs/20230818-Dropout/README.md:21]: @mgouicem , thanks!
@igorsafo , I updated README

File: rfcs/20230818-Dropout/README.md (hunk 4 above)
Comment (itaraban) [reply rfcs/20230818-Dropout/README.md]: @igorsafo , thanks for you suggestion!
I updated README and PoC

File: rfcs/20230818-Dropout/README.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    const_dnnl_memory_desc_t cdesc;
+    uint8_t enabled_u8;
+    error::wrap_c_api(
+            dnnl_primitive_attr_get_dropout(get(), &enabled_u8, &cdesc),
+            "could not get parameters of a dropout attribute");
+    dnnl_memory_desc_t cloned_md = nullptr;
+    error::wrap_c_api(dnnl_memory_desc_clone(&cloned_md, cdesc),
+            "could not clone a memory descriptor");
+    mask_desc = memory::desc(cloned_md);
+    enabled = enabled_u8;
```
Comment (igorsafo) [inline rfcs/20230818-Dropout/README.md:49]:
```suggestion
    error::wrap_c_api(
            dnnl_primitive_attr_get_dropout(get(), &cdesc),
            "could not get parameters of a dropout attribute");
    dnnl_memory_desc_t cloned_md = nullptr;
    error::wrap_c_api(dnnl_memory_desc_clone(&cloned_md, cdesc),
            "could not clone a memory descriptor");
    mask_desc = memory::desc(cloned_md);
```

Comment (itaraban) [reply rfcs/20230818-Dropout/README.md]: Thanks, changed

**Review by dzarukin — APPROVED** [review]

**Comment by dzarukin** [comment]
Merged as functionality is implemented in the library.
